src/tools/non-promotion-reasons.md

```diff
@@ -1,45 +1,399 @@
+the right-hand side of the assignment,
+similar to what happens in loops.
+
+***[PENDING: what's the fix?]***
+
+
+### Subtype mismatch
+
+**The cause:**
+The type you're trying to promote to isn't a subtype of
+the variable's current promoted type
+(or wasn't a subtype at the time of the promotion attempt).
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(Object o) {
+  if (o is Comparable) {             // (1)
+    if (o is Pattern) {              // (2)
+      print(o.matchAsPrefix('foo')); // (3) ERROR
+    }
+  }
+}
+{% endprettify %}
+
+In this example, `o` is promoted to `Comparable` at (1), but
+it isn't promoted to `Pattern` at (2),
+because `Pattern` isn't a subtype of `Comparable`.
+(The rationale is that if it did promote,
+then you wouldn't be able to use methods on `Comparable`.)
+Note that just because `Pattern` isn't a subtype of `Comparable`
+doesn't mean the code at (3) is dead;
+`o` might have a type — like `String` —
+that implements both `Comparable` and `Pattern`.
+
+***[PENDING: what's the fix?]***
+
+
+### Write captured by a local function
+
+**The cause:**
+The variable has been write captured by
+a local function or function expression.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  if (i == null) return;
+  var foo = () {
+    i = j;
+  };
+  print(i.isEven); // ERROR
+}
+{% endprettify %}
+
+Flow analysis reasons that as soon as the definition of `foo` is reached,
+it might get called at any time,
+therefore it's no longer safe to promote `i` at all.
+As with loops, this demotion happens regardless of
+the type of the right hand side of the assignment.
+
+***[PENDING: what's the fix?]***
+
+
+### Written outside of the current closure or function expression {#write-outer}
+
+**The cause:**
+The variable is written to outside of a closure or function expression,
+and the type promotion location is
+inside the closure or function expression.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  if (i == null) return;
+  var foo = () {
+    print(i.isEven); // (1) ERROR
+  };
+  i = j;             // (2)
+}
+{% endprettify %}
+
+Flow analysis reasons that there's no way to determine
+when `foo` might get called,
+so it might get called after the assignment at (2),
+and thus the promotion might no longer be valid.
+As with loops, this demotion happens regardless of the type of
+the right hand side of the assignment.
+
+A particularly nasty case looks like this:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i) {
+  i ??= 0;
+  var foo = () {
+    print(i.isEven); // ERROR
+  };
+}
+{% endprettify %}
+
+In this case, a human can see that the promotion is safe because
+the only write to `i` uses a non-null value and
+happens before `foo` is ever created.
+But [flow analysis isn't that smart][1536].
+
+[1536]: https://github.com/dart-lang/language/issues/1536
+
+***[PENDING: what's the fix?]***
```

Comment:
The best fix for this "particularly nasty" case is probably to make a fresh variable, e.g.:
```dart
void f(int? i) {
  var j = i ?? 0;
  var foo = () {
    print(j.isEven); // OK
  };
}
```
This works because `i ?? 0` has type non-nullable `int`; this will be used as the inferred type of `j`, so regardless of whether there's an assignment to `j` later, it can never take on a non-null value.